packages/loopring_v3/contracts/lib/SignatureUtil.sol

```diff
@@ -155,15 +158,24 @@ library SignatureUtil
         SignatureType signatureType = SignatureType(signature.toUint8(signatureTypeOffset));
 
         bytes memory stripped = signature.slice(0, signatureTypeOffset);
-        bytes32 hash = getDataHash(data);
 
         if (signatureType == SignatureType.EIP_712) {
-            return recoverECDSASigner(hash, stripped) == signer;
+            (address addr1, address addr2) = recoverECDSASigner(data, stripped);
+            return addr1 == signer || addr2 == signer;
         } else if (signatureType == SignatureType.ETH_SIGN) {
-            hash = keccak256(
-                abi.encodePacked("\x19Ethereum Signed Message:\n32", hash)
-            );
-            return recoverECDSASigner(hash, stripped) == signer;
+            if (data.length == 32) {
+                bytes32 hash = keccak256(
+                    abi.encodePacked("\x19Ethereum Signed Message:\n32", data.toBytes32(0))
+                );
+                if (recoverECDSASigner(hash, stripped) == signer) {
+                    return true;
+                }
+            } else {
+                bytes32 hash = keccak256(
+                    abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(data))
+                );
+                return recoverECDSASigner(hash, stripped) == signer;
+            }
```

Comment:
Will this work as expected? Seems like this will still have the old behaviour where it is assumed bytes of length 32 are already hashed. Totally not sure, it's a bit hard to keep track. :)

Reply:
Yes, it works. Basically it does this:
- if the data is 32 byte-long, try to treat it as hash, if it fails, try to use as the raw data of 32 bytes.
- if the data is not 32 long, use it as raw data.

Reply:
> if the data is 32 byte-long, try to treat it as hash, if it fails, try to use as the raw data of 32 bytes.

If the data is 32 bytes long, and `recoverECDSASigner(hash, stripped) == signer` is false, it won't get into the `else` no? So it won't try with `keccak256(data)`?

Reply:
Sharp eyes. I fixed the issue by removing an "else".